### PR TITLE
Fix #5037 - Adjust log formatters for workflow (stdout + run.log)

### LINKED ISSUE #5037: Adjust LogSink log formatter
Issue overview
--------------

Prompted by the fact that the run.log looks quite different between the classic (Ruby) and the C++ CLI.

Additionally, the default formatter is not great. The use of Integers for the log level is unnecessarily confusing. I think I know how to turn the LogLevel enum into a string (eg: "Debug" instead of `-1`)

https://github.com/NREL/OpenStudio/blob/3f35a4ce446e1e91a4ec0c559aeca226b2715099/src/utilities/core/LogSink.cpp#L126-L131

## Current Behavior

**C++**

```
[openstudio.measure.OSRunner] <0> runner.registerWarning called
[openstudio.workflow.OSWorkflow] <0> No results for objective function IsNonExisting.NonExisting
```

**Ruby**:

```
[13:54:18.023960 WARN] runner.registerWarning called
[13:54:18.056100 WARN] No results for objective function IsNonExisting.NonExisting
```

## Steps to Reproduce

1.
2.
3.
4.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version):
 - Version of OpenStudio (if using an intermediate build, include SHA):

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #5037

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/workflow/OSWorkflow.cpp [hunk 1]
```diff
@@ -216,21 +216,44 @@ void OSWorkflow::saveIDFToRootDirIfDebug() {
...
 
+  openstudio::Logger::instance().addTimeStampToLogger();  // Needed for run.log formatting
+  openstudio::Logger::instance().standardOutLogger().setFormatter(&standardFormatterWithStringSeverity);
+
+  // TODO: ideally we want stdErr logger to always receive Error and Fatal
+  // and stdOut logger should receive all the others. This is definitely doable (cf LogSink::updateFilter) but now is not the time.
   if (!m_show_stdout) {
-    openstudio::Logger::instance().standardOutLogger().disable();
-  } else if (workflowJSON.runOptions()->debug()) {
+    openstudio::Logger::instance().standardOutLogger().setLogLevel(Error);  // Still show errors
```
Comment (jmarrec) [inline src/workflow/OSWorkflow.cpp:255]: If `--show-stdout` not passed, still show Error and Fatals + leave a TODO to clean up later.

File: src/workflow/OSWorkflow.cpp [hunk 2]
```diff
@@ -216,21 +216,44 @@ void OSWorkflow::saveIDFToRootDirIfDebug() {
...
-    oriLogLevel = *l_;
-  }
+  LogLevel oriLogLevel = openstudio::Logger::instance().standardOutLogger().logLevel().value_or(Warn);
   LogLevel targetLogLevel = oriLogLevel;
   if (workflowJSON.runOptions()->debug() && oriLogLevel > Debug) {
     targetLogLevel = Debug;
   }
 
+  openstudio::Logger::instance().addTimeStampToLogger();  // Needed for run.log formatting
+  openstudio::Logger::instance().standardOutLogger().setFormatter(&standardFormatterWithStringSeverity);
```
Comment (jmarrec) [inline src/workflow/OSWorkflow.cpp:250]: Stdout: `[Channel] <Debug> MSG`

File: src/workflow/OSWorkflow.cpp [hunk 3]
```diff
@@ -252,7 +275,11 @@ bool OSWorkflow::run() {
   if (!openstudio::filesystem::is_directory(runDirPath)) {
     openstudio::filesystem::create_directory(runDirPath);
   }
+
   FileLogSink logFile(runDirPath / "run.log");
+  constexpr bool use_workflow_gem_fmt = true;
+  constexpr bool include_channel = true;  // or workflowJSON.runOptions()->debug();
+  logFile.useWorkflowGemFormatter(use_workflow_gem_fmt, include_channel);
```
Comment (jmarrec) [inline src/workflow/OSWorkflow.cpp:282]: Run.log: `[11:12:13.123456 DEBUG] [Channel] MSG`

Same as the Old Ruby CLI, but with the channel in addition

**Comment by jmarrec** [comment]
Example:

* When show stdout not passed, still show error.
* See new formats!

```
$ $os_build2/Products/openstudio run --postprocess_only -w reportingmeasure.osw
[openstudio.measure.OSRunner] <Error> runner.registerError called
[openstudio.workflow.OSWorkflow] <Error> Found error in state 'ReportingMeasures' with message: 'Measure 'ReportingMeasureRegistersError' reported an error with [runner.registerError called]'
Failed to run workflow. Last Error:
  Found error in state 'ReportingMeasures' with message: 'Measure 'ReportingMeasureRegistersError' reported an error with [runner.registerError called]'

$ cat run/run.log
[18:38:06.159800 WARN] [openstudio.EpwFile] Header start day of the week and actual start day of the week do not match in EPW file '/Users/julien/Software/Others/OS-build2/resources/Examples/compact_osw/files/srrl_2013_amy.epw'. Data will be treated as typical (TMY)
[18:38:06.936419 ERROR] [openstudio.measure.OSRunner] runner.registerError called
[18:38:06.936646 ERROR] [openstudio.workflow.OSWorkflow] Found error in state 'ReportingMeasures' with message: 'Measure 'ReportingMeasureRegistersError' reported an error with [runner.registerError called]'
```

**Review by kbenne — APPROVED** [review]

**Pull request merged**